kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,399 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteBucket stores the bucket id and the buckets transaction.
+type readWriteBucket struct {
+	// id is used to identify the bucket. If id is null, it refers to the
+	// root bucket.
+	id *int64
+
+	// tx holds the parent transaction.
+	tx *readWriteTx
+
+	table string
+}
+
+// newReadWriteBucket creates a new rw bucket with the passed transaction
+// and bucket id.
+func newReadWriteBucket(tx *readWriteTx, table string, id *int64) *readWriteBucket {
+	return &readWriteBucket{
+		id:    id,
+		tx:    tx,
+		table: table,
+	}
+}
+
+// NestedReadBucket retrieves a nested read bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadBucket(key []byte) walletdb.ReadBucket {
+	return b.NestedReadWriteBucket(key)
+}
+
+func parentSelector(id *int64) string {
+	if id == nil {
+		return "parent_id IS NULL"
+	}
+	return fmt.Sprintf("parent_id=%v", *id)
+}
+
+// ForEach invokes the passed function with every key/value pair in
+// the bucket. This includes nested buckets, in which case the value
+// is nil, but it does not include the key/value pairs within those
+// nested buckets.
+func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
+	rows, err := b.tx.tx.QueryContext(
+		context.TODO(),
+		"SELECT key,value FROM "+b.table+" WHERE "+parentSelector(b.id)+" ORDER BY key",
+	)
+	if err != nil {
+		return err
+	}
+
+	type kv struct {
+		k, v []byte
+	}
+
+	// Read all data in memory, to unblock the connection (nested queries).
+	//
+	// TODO: Convert to paginated reads to prevent OOM on large tables.
+	var kvs []kv
+	for rows.Next() {
+		var (
+			key   []byte
+			value []byte
+		)
+		err := rows.Scan(&key, &value)
+		if err != nil {
+			rows.Close()
+			return err
+		}
+
+		kvs = append(kvs, kv{k: key, v: value})
+	}
+	rows.Close()
+
+	for _, kv := range kvs {
+		err = cb(kv.k, kv.v)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get returns the value for the given key. Returns nil if the key does
+// not exist in this bucket.
+func (b *readWriteBucket) Get(key []byte) []byte {
+	// Return nil if the key is empty.
+	if len(key) == 0 {
+		return nil
+	}
+
+	var value *[]byte
+	err := b.tx.tx.QueryRowContext(
+		context.TODO(),
```

Comment:
I didn't do that just yet, because holding a context in an object is an anti-pattern. But as a temporary solution we indeed may have no other option to be able to abort the application.

Reply:
Fixed